Error when adding visual testing plugin

```
$ ./gradlew build eclipse

2019-09-06 16:38:41,573 [main] ERROR org.vividus.runner.VividusInitializationChecker - org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'visualSteps': Unsatisfied dependency expressed through field 'visualTestingEngine'; nested exception is org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'visualTestingEngine': Unsatisfied dependency expressed through field 'baselineRepository'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'fileSystemBaselineRepository' defined in URL [jar:file:/D:/***/.out/caches/modules-2/files-2.1/org.vividus/vividus-plugin-visual/0.1.0-SNAPSHOT/f040643f48b9fe19b27a5c474f160f8a3e53c07/vividus-plugin-visual-0.1.0-SNAPSHOT.jar!/spring.xml]: Invocation of init method failed; nested exception is java.lang.IllegalArgumentException: Resource with name /baselines for class org.vividus.visual.engine.FileSystemBaselineRepository is not found
2019-09-06 16:38:41,579 [main] ERROR org.vividus.runner.VividusInitializationChecker - org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'visualTestingEngine': Unsatisfied dependency expressed through field 'baselineRepository'; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'fileSystemBaselineRepository' defined in URL [jar:file:/D:/***/.out/caches/modules-2/files-2.1/org.vividus/vividus-plugin-visual/0.1.0-SNAPSHOT/f040643f48b9fe19b27a5c474f160f8a3e53c07/vividus-plugin-visual-0.1.0-SNAPSHOT.jar!/spring.xml]: Invocation of init method failed; nested exception is java.lang.IllegalArgumentException: Resource with name /baselines for class org.vividus.visual.engine.FileSystemBaselineRepository is not found
2019-09-06 16:38:41,583 [main] ERROR org.vividus.runner.VividusInitializationChecker - org.springframework.beans.factory.BeanCreationE                               xception: Error creating bean with name 'fileSystemBaselineRepository' defined in URL [jar:file:/D:/***/.out/caches/modules-2/files-2.1/org.vividus/vividus-plugin-visual/0.1.0-SNAPSHOT/f040643f48b9fe19b27a5c474f160f8a3e53c07/vividus-plugin-visual-0.1.0-SNAPSHOT.jar!/spring.xml]: Invocation of init method failed; nested exception is java.lang.IllegalArgumentException: Resource with name /baselines for class org.vividus.visual.engine.FileSystemBaselineRepository is not found
2019-09-06 16:38:43,017 [main] WARN  org.apache.tika.parser.SQLite3Parser - org.xerial's sqlite-jdbc is not loaded.
Please provide the jar on your classpath to parse sqlite files.
See tika-parsers/pom.xml for the correct version.
Exception in thread "main" java.lang.RuntimeException: Initialization of beans has been failed
        at org.vividus.runner.VividusInitializationChecker.main(VividusInitializationChecker.java:86)

> Task :testVividusInitialization FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':testVividusInitialization'.
> Process 'command 'C:\Program Files\Java\jdk-11.0.1\bin\java.exe'' finished with non-zero exit value 1

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 21s
9 actionable tasks: 3 executed, 6 up-to-date

```

**Workaround:** Adding of /baselines folder in src/main/resources fixes the build and vis plugin added successfully.

**Suggestion:** To fix this baseline folder login inside the framework to avoid issues.

--- a/vividus-plugin-visual/src/main/java/org/vividus/visual/engine/FileSystemBaselineRepository.java
+++ b/vividus-plugin-visual/src/main/java/org/vividus/visual/engine/FileSystemBaselineRepository.java
@@ -19,9 +19,12 @@ package org.vividus.visual.engine;
 import java.io.File;
 import java.io.IOException;
 import java.util.Optional;
+import java.util.function.Supplier;
 
 import javax.imageio.IIOException;
 import javax.imageio.ImageIO;
+
+import com.google.common.base.Suppliers;
 
 import org.apache.tools.ant.util.StringUtils;
 import org.slf4j.Logger;
@@ -38,22 +41,22 @@ public class FileSystemBaselineRepository implements IBaselineRepository
 
     private File baselinesFolder;
 
-    public void init()
-    {
+    private final Supplier<File> baselineFolderResolver = Suppliers.memoize(() -> {
         if (!baselinesFolder.isAbsolute())
         {
             String replacement = "/";
-            baselinesFolder = ResourceUtils.loadFile(FileSystemBaselineRepository.class,
+            return ResourceUtils.loadFile(FileSystemBaselineRepository.class,
                             StringUtils.removePrefix(baselinesFolder.toString(), ".").replace("\\", replacement));
         }
-    }
+        return baselinesFolder;
+    });
 
     @Override
     public Optional<Screenshot> getBaseline(String baselineName) throws IOException
     {
         try
         {
-            File baselineFile = new File(baselinesFolder, appendExtension(baselineName));
+            File baselineFile = new File(baselineFolderResolver.get(), appendExtension(baselineName));
             Optional<Screenshot> loadedScreenshot = Optional
                     .ofNullable(ImageIO.read(baselineFile))
                     .map(Screenshot::new);
@@ -77,7 +80,7 @@ public class FileSystemBaselineRepository implements IBaselineRepository
     @Override
     public void saveBaseline(Screenshot toSave, String baselineName) throws IOException
     {
-        File baselineToSave = new File(baselinesFolder, baselineName);
+        File baselineToSave = new File(baselineFolderResolver.get(), baselineName);
         ImageUtils.writeAsPng(toSave.getImage(), baselineToSave);
         LOGGER.info("Baseline saved to: {}", appendExtension(baselineToSave.getAbsolutePath()));
     }

--- a/vividus-plugin-visual/src/main/resources/spring.xml
+++ b/vividus-plugin-visual/src/main/resources/spring.xml
@@ -18,8 +18,7 @@
         <property name="overrideBaselines" value="${web.visual.override-baselines}" />
     </bean>
 
-    <bean id="fileSystemBaselineRepository" class="org.vividus.visual.engine.FileSystemBaselineRepository"
-         init-method="init">
+    <bean id="fileSystemBaselineRepository" class="org.vividus.visual.engine.FileSystemBaselineRepository">
         <property name="baselinesFolder" value="${web.visual.baselines-folder}" />
     </bean>
 

--- a/vividus-plugin-visual/src/test/java/org/vividus/visual/engine/FileSystemBaselineRepositoryTests.java
+++ b/vividus-plugin-visual/src/test/java/org/vividus/visual/engine/FileSystemBaselineRepositoryTests.java
@@ -53,7 +53,7 @@ import ru.yandex.qatools.ashot.util.ImageTool;
 @SuppressFBWarnings("DMI_HARDCODED_ABSOLUTE_FILENAME")
 class FileSystemBaselineRepositoryTests
 {
-    private static final String BASELINE = "BASELINE";
+    private static final String BASELINE = "baseline";
 
     private static final File BASELINES_FOLDER = new File("./baselines");
 
@@ -65,8 +65,7 @@ class FileSystemBaselineRepositoryTests
     void shouldLoadBaselineFromFileSystem() throws IOException
     {
         fileSystemBaselineRepository.setBaselinesFolder(BASELINES_FOLDER);
-        fileSystemBaselineRepository.init();
-        assertThat(fileSystemBaselineRepository.getBaseline("baseline").get().getImage(),
+        assertThat(fileSystemBaselineRepository.getBaseline(BASELINE).get().getImage(),
                 ImageTool.equalImage(baseline));
     }
 
@@ -87,7 +86,6 @@ class FileSystemBaselineRepositoryTests
     void shouldReturnEmptyImageForMissingBaseline() throws IOException
     {
         fileSystemBaselineRepository.setBaselinesFolder(BASELINES_FOLDER);
-        fileSystemBaselineRepository.init();
         assertEquals(fileSystemBaselineRepository.getBaseline("missing_baseline"), Optional.empty());
     }
 
@@ -95,14 +93,13 @@ class FileSystemBaselineRepositoryTests
     void shouldThrowIllegalArgumentExceptionForNotExistingFoler()
     {
         fileSystemBaselineRepository.setBaselinesFolder(new File("no_such_folder"));
-        assertThrows(IllegalArgumentException.class, fileSystemBaselineRepository::init);
+        assertThrows(IllegalArgumentException.class, () -> fileSystemBaselineRepository.getBaseline(BASELINE));
     }
 
     @Test
     void shouldThrowExceptionWhenBaselineNotLoaded()
     {
         fileSystemBaselineRepository.setBaselinesFolder(BASELINES_FOLDER);
-        fileSystemBaselineRepository.init();
         ResourceLoadException exception = assertThrows(ResourceLoadException.class,
             () -> fileSystemBaselineRepository.getBaseline("corrupted_image").get().getImage());
         assertThat(exception.getMessage(), Matchers.matchesRegex(
@@ -113,7 +110,6 @@ class FileSystemBaselineRepositoryTests
     void shouldSaveBaselineIntoFolder(@TempDir File folder) throws IOException
     {
         fileSystemBaselineRepository.setBaselinesFolder(folder);
-        fileSystemBaselineRepository.init();
         Screenshot screenshot = mock(Screenshot.class);
         when(screenshot.getImage()).thenReturn(baseline);
         fileSystemBaselineRepository.saveBaseline(screenshot, BASELINE);

